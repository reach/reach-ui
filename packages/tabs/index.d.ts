declare module "@reach/tabs" {
  import * as React from "react";

  export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

  export type TabsProps = {
    children: React.ReactNode;
    as?: string;
    index?: number;
    readOnly?: boolean;
    defaultIndex?: number;
    onChange?: (index: number) => void;
  } & Omit<React.HTMLProps<HTMLDivElement>, "onChange">;

  export type TabContainerProps = {
    children?: React.ReactNode;
    selectedIndex?: number;
    as?: string;
  } & React.HTMLProps<HTMLElement>;

  export type TabPanelProps = {
    children?: React.ReactNode;
    isSelected?: boolean;
    as?: string;
  } & React.HTMLProps<HTMLElement>;

  export type TabProps = {
    disabled?: boolean;
  } & TabPanelProps;

  export const Tabs: React.FunctionComponent<TabsProps>;
  export const TabList: React.FunctionComponent<TabContainerProps>;
  export const TabPanels: React.FunctionComponent<TabContainerProps>;
  export const Tab: React.FunctionComponent<TabProps>;
  export const TabPanel: React.FunctionComponent<TabPanelProps>;
}
