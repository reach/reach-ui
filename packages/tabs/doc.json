[
  {
    "description": "React component.",
    "displayName": "ComponentTest",
    "methods": [],
    "props": {
      "children": {
        "type": { "name": "node" },
        "required": true,
        "description": "Does a thing"
      }
    }
  }
]
