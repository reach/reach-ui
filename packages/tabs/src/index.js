import React, {
  cloneElement,
  useState,
  useEffect,
  useRef,
  forwardRef
} from "react";
import { node, func, number } from "prop-types";
import warning from "warning";
import { wrapEvent } from "@reach/utils";
import { useId } from "@reach/auto-id";

////////////////////////////////////////////////////////////////////////////////
export const Tabs = forwardRef(function Tabs(
  {
    children,
    as: Comp = "div",
    onChange,
    index: controlledIndex = undefined,
    readOnly = false,
    defaultIndex,
    ...props
  },
  ref
) {
  // useRef because you shouldn't switch between controlled/uncontrolled
  const { current: isControlled } = useRef(controlledIndex != null);

  warning(
    !(isControlled && controlledIndex == null),
    "Tabs is changing from controlled to uncontrolled. Tabs should not switch from controlled to uncontrolled (or vice versa). Decide between using a controlled or uncontrolled Tabs for the lifetime of the component. Check the `index` prop being passed in."
  );

  warning(
    !(!isControlled && controlledIndex != null),
    "Tabs is changing from uncontrolled to controlled. Tabs should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled Tabs for the lifetime of the component. Check the `index` prop being passed in."
  );

  const _id = useId();

  // we only manage focus if the user caused the update vs.
  // a new controlled index coming in
  const _userInteractedRef = useRef(false);

  const _selectedPanelRef = useRef(null);

  const [selectedIndex, setSelectedIndex] = useState(defaultIndex || 0);

  const clones = React.Children.map(children, child => {
    // ignore random <div/>s etc.
    if (!child || typeof child.type === "string") return child;
    return cloneElement(child, {
      selectedIndex: isControlled ? controlledIndex : selectedIndex,
      _id,
      _userInteractedRef,
      _selectedPanelRef,
      _onFocusPanel: () =>
        _selectedPanelRef.current && _selectedPanelRef.current.focus(),
      _onSelectTab: readOnly
        ? () => {}
        : index => {
            _userInteractedRef.current = true;
            onChange && onChange(index);
            if (!isControlled) {
              setSelectedIndex(index);
            }
          }
    });
  });

  return <Comp data-reach-tabs="" ref={ref} {...props} children={clones} />;
});

if (__DEV__) {
  Tabs.propTypes = {
    children: node.isRequired,
    onChange: func,
    index: (props, name, compName, ...rest) => {
      if (
        props.index > -1 &&
        props.onChange == null &&
        props.readOnly !== true
      ) {
        return new Error(
          "You provided a `value` prop to `Tabs` without an `onChange` handler. This will render a read-only tabs element. If the tabs should be mutable use `defaultIndex`. Otherwise, set `onChange`."
        );
      } else {
        return number(name, props, compName, ...rest);
      }
    },
    defaultIndex: number
  };
}

////////////////////////////////////////////////////////////////////////////////
export const TabList = forwardRef(function TabList(
  { children, as: Comp = "div", onKeyDown, ...clonedProps },
  ref
) {
  const {
    selectedIndex,
    _onSelectTab,
    _userInteractedRef,
    _onFocusPanel,
    _selectedPanelRef,
    _id,
    ...htmlProps
  } = clonedProps;

  const clones = React.Children.map(children, (child, index) => {
    return cloneElement(child, {
      isSelected: index === selectedIndex,
      _id: makeId(_id, index),
      _userInteractedRef,
      _onSelect: () => _onSelectTab(index)
    });
  });

  const handleKeyDown = wrapEvent(onKeyDown, event => {
    const enabledIndexes = React.Children.map(children, (child, index) =>
      child.props.disabled === true ? null : index
    ).filter(index => index != null); // looks something like: [0, 2, 3, 5]
    const enabledSelectedIndex = enabledIndexes.indexOf(selectedIndex);

    switch (event.key) {
      case "ArrowRight": {
        const nextEnabledIndex =
          (enabledSelectedIndex + 1) % enabledIndexes.length;
        const nextIndex = enabledIndexes[nextEnabledIndex];
        _onSelectTab(nextIndex);
        break;
      }
      case "ArrowLeft": {
        const count = enabledIndexes.length;
        const nextEnabledIndex = (enabledSelectedIndex - 1 + count) % count;
        const nextIndex = enabledIndexes[nextEnabledIndex];
        _onSelectTab(nextIndex);
        break;
      }
      case "ArrowDown": {
        // don't scroll down
        event.preventDefault();
        _onFocusPanel();
        break;
      }
      case "Home": {
        _onSelectTab(0);
        break;
      }
      case "End": {
        _onSelectTab(React.Children.count(children) - 1);
        break;
      }
      default: {
      }
    }
  });

  return (
    <Comp
      data-reach-tab-list=""
      ref={ref}
      role="tablist"
      onKeyDown={handleKeyDown}
      children={clones}
      {...htmlProps}
    />
  );
});

if (__DEV__) {
  TabList.propTypes = {
    children: node
  };
}

////////////////////////////////////////////////////////////////////////////////
export const Tab = forwardRef(function Tab(
  { children, as: Comp = "button", ...rest },
  forwardedRef
) {
  const { isSelected, _userInteractedRef, _onSelect, _id, ...htmlProps } = rest;
  const htmlType =
    Comp === "button" && htmlProps.type == null ? "button" : undefined;

  const ownRef = useRef(null);
  const ref = forwardedRef || ownRef;

  useUpdateEffect(() => {
    if (isSelected && ref.current && _userInteractedRef.current) {
      _userInteractedRef.current = false;
      ref.current.focus();
    }
  }, [isSelected]);

  return (
    <Comp
      data-reach-tab=""
      ref={ref}
      role="tab"
      id={makeId("tab", _id)}
      tabIndex={isSelected ? 0 : -1}
      aria-selected={isSelected}
      aria-controls={makeId("panel", _id)}
      data-selected={isSelected ? "" : undefined}
      onClick={_onSelect}
      children={children}
      type={htmlType}
      {...htmlProps}
    />
  );
});

if (__DEV__) {
  Tab.propTypes = {
    children: node
  };
}

////////////////////////////////////////////////////////////////////////////////
export const TabPanels = forwardRef(function TabPanels(
  { children, as: Comp = "div", ...rest },
  forwardedRef
) {
  const {
    selectedIndex,
    _selectedPanelRef,
    _userInteractedRef,
    _onFocusPanel,
    _onSelectTab,
    _id,
    ...htmlAttrs
  } = rest;

  const clones = React.Children.map(children, (child, index) =>
    cloneElement(child, {
      isSelected: index === selectedIndex,
      _selectedPanelRef,
      _id: makeId(_id, index)
    })
  );

  return (
    <Comp
      data-reach-tab-panels=""
      ref={forwardedRef}
      {...htmlAttrs}
      children={clones}
    />
  );
});

if (__DEV__) {
  TabPanels.propTypes = {
    children: node
  };
}

////////////////////////////////////////////////////////////////////////////////
export const TabPanel = forwardRef(function TabPanel(
  { children, as: Comp = "div", ...rest },
  // TODO: Do we need to use this ref?
  forwardedRef
) {
  const { isSelected, _selectedPanelRef, _id, ...htmlProps } = rest;

  return (
    <Comp
      data-reach-tab-panel=""
      ref={isSelected ? _selectedPanelRef : undefined}
      role="tabpanel"
      tabIndex={-1}
      aria-labelledby={makeId("tab", _id)}
      hidden={!isSelected}
      id={makeId("panel", _id)}
      children={children}
      {...htmlProps}
    />
  );
});

if (__DEV__) {
  TabPanel.propTypes = {
    children: node
  };
}

////////////////////////////////////////////////////////////////////////////////
// TODO: move into @reach/utils when something else needs it
function useUpdateEffect(effect, deps) {
  const mounted = useRef(false);
  useEffect(() => {
    if (mounted.current) {
      effect();
    } else {
      mounted.current = true;
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, deps);
}

const makeId = (id, index) => `${id}--${index}`;
