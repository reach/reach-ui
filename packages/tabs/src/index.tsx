/**
 * Welcom to @reach/tabs!
 *
 * An accessible tabs component.
 *
 * The `Tab` and `TabPanel` elements are associated by their order in the tree.
 * None of the components are empty wrappers, each is associated with a real DOM
 * element in the document, giving you maximum control over styling and composition.
 *
 * You can render any other elements you want inside of `Tabs`, but `TabList`
 * should only render `Tab` elements, and `TabPanels` should only render
 * `TabPanel` elements.
 *
 * TODO: Consider manual tab activation
 * https://www.w3.org/TR/wai-aria-practices-1.1/examples/tabs/tabs-2/tabs.html
 *
 * TODO: Consider `orientation` prop to account for keyboard behavior
 *       - horizontal-top
 *       - horizontal-bottm
 *       - vertical-left
 *       - vertical-right
 *
 * @see Docs     https://reacttraining.com/reach-ui/tabs
 * @see Source   https://github.com/reach/reach-ui/tree/master/packages/tabs
 * @see WAI-ARIA https://www.w3.org/TR/wai-aria-practices-1.1/#tabs
 */

import React, {
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState,
  Children,
  useCallback
} from "react";
import PropTypes from "prop-types";
import warning from "warning";
import {
  boolOrBoolString,
  checkStyles,
  cloneValidElement,
  createDescendantContext,
  createNamedContext,
  DescendantProvider,
  forwardRefWithAs,
  makeId,
  noop,
  useDescendant,
  useDescendants,
  useForkedRef,
  useIsomorphicLayoutEffect as useLayoutEffect,
  useUpdateEffect,
  wrapEvent
} from "@reach/utils";
import { useId } from "@reach/auto-id";

interface ITabsContext {
  id: string;
  isControlled: boolean;
  onFocusPanel: () => void;
  onSelectTab: (index: number) => void;
  selectedIndex: number;
  selectedPanelRef: React.MutableRefObject<HTMLElement | null>;
  setSelectedIndex: React.Dispatch<React.SetStateAction<number>>;
  userInteractedRef: React.MutableRefObject<boolean>;
}

const TabsDescendantsContext = createDescendantContext<HTMLElement>(
  "TabsDescendantsContext"
);

const TabPanelDescendantsContext = createDescendantContext<HTMLElement>(
  "TabPanelDescendantsContext"
);
const TabsContext = createNamedContext("TabsContext", {} as ITabsContext);
const useTabsContext = () => useContext(TabsContext);

////////////////////////////////////////////////////////////////////////////////

/**
 * Tabs
 *
 * The parent component of the tab interface.
 *
 * @see Docs https://reacttraining.com/reach-ui/tabs#tabs
 */
export const Tabs = forwardRefWithAs<TabsProps, "div">(function Tabs(
  {
    as: Comp = "div",
    children,
    defaultIndex,
    index: controlledIndex = undefined,
    onChange,
    readOnly = false,
    ...props
  },
  ref
) {
  let isControlled = useRef(controlledIndex != null);
  useEffect(() => {
    if (__DEV__) {
      warning(
        !(isControlled.current && controlledIndex == null),
        "Tabs is changing from controlled to uncontrolled. Tabs should not switch from controlled to uncontrolled (or vice versa). Decide between using a controlled or uncontrolled Tabs for the lifetime of the component. Check the `index` prop being passed in."
      );
      warning(
        !(!isControlled.current && controlledIndex != null),
        "Tabs is changing from uncontrolled to controlled. Tabs should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled Tabs for the lifetime of the component. Check the `index` prop being passed in."
      );
    }
  }, [controlledIndex]);

  let _id = useId(props.id);
  let id = props.id ?? makeId("tabs", _id);

  /*
   * We only manage focus if the user caused the update vs. a new controlled
   * index coming in.
   */
  let userInteractedRef = useRef(false);

  let selectedPanelRef = useRef<HTMLElement | null>(null);
  let [selectedIndex, setSelectedIndex] = useState(defaultIndex || 0);
  let [tabs, setTabs] = useDescendants<HTMLElement>();

  const context: ITabsContext = useMemo(() => {
    return {
      isControlled: isControlled.current,
      selectedIndex: isControlled.current
        ? (controlledIndex as number)
        : selectedIndex,
      id,
      userInteractedRef,
      selectedPanelRef,
      setSelectedIndex: isControlled.current ? noop : setSelectedIndex,
      onFocusPanel: () => {
        selectedPanelRef.current?.focus();
      },
      onSelectTab: readOnly
        ? noop
        : (index: number) => {
            userInteractedRef.current = true;
            onChange && onChange(index);
            if (!isControlled.current) {
              setSelectedIndex(index);
            }
          }
    };
  }, [controlledIndex, id, onChange, readOnly, selectedIndex]);

  useEffect(() => checkStyles("tabs"), []);

  return (
    <DescendantProvider
      context={TabsDescendantsContext}
      items={tabs}
      set={setTabs}
    >
      <TabsContext.Provider value={context}>
        <Comp {...props} ref={ref} data-reach-tabs="" id={props.id}>
          {children}
        </Comp>
      </TabsContext.Provider>
    </DescendantProvider>
  );
});

/**
 * @see Docs https://reacttraining.com/reach-ui/tabs#tabs-props
 */
export type TabsProps = {
  /**
   * Tabs expects `<TabList>` and `<TabPanels>` as children. The order doesn't
   * matter, you can have tabs on the top or the bottom. In fact, you could have
   * tabs on both the bottom and the top at the same time. You can have random
   * elements inside as well.
   *
   * @see Docs https://reacttraining.com/reach-ui/tabs#tabs-props
   */
  children: React.ReactNode;
  /**
   * Like form inputs, a tab's state can be controlled by the owner. Make sure
   * to include an `onChange` as well, or else the tabs will not be interactive.
   *
   * @see Docs https://reacttraining.com/reach-ui/tabs#tabs-props
   */
  index?: number;
  /**
   * @see Docs https://reacttraining.com/reach-ui/tabs#tabs-props
   */
  readOnly?: boolean;
  /**
   * Starts the tabs at a specific index.
   *
   * @see Docs https://reacttraining.com/reach-ui/tabs#tabs-props
   */
  defaultIndex?: number;
  /**
   * Calls back with the tab index whenever the user changes tabs, allowing your
   * app to synchronize with it.
   *
   * @see Docs https://reacttraining.com/reach-ui/tabs#tabs-props
   */
  onChange?: (index: number) => void;
};

if (__DEV__) {
  Tabs.displayName = "Tabs";
  Tabs.propTypes = {
    children: PropTypes.node.isRequired,
    onChange: PropTypes.func,
    index: (props, name, compName, location, propName) => {
      let val = props[name];
      if (
        props.index > -1 &&
        props.onChange == null &&
        props.readOnly !== true
      ) {
        return new Error(
          "You provided a value prop to `" +
            compName +
            "` without an `onChange` handler. This will render a read-only tabs element. If the tabs should be mutable use `defaultIndex`. Otherwise, set `onChange`."
        );
      } else if (props[name] != null && typeof props[name] !== "number") {
        return new Error(
          `Invalid prop \`${propName}\` supplied to \`${compName}\`. Expected \`number\`, received \`${
            Array.isArray(val) ? "array" : typeof val
          }\`.`
        );
      }
      return null;
    },
    defaultIndex: PropTypes.number
  };
}

////////////////////////////////////////////////////////////////////////////////

/**
 * TabList
 *
 * The parent component of the tabs.
 *
 * @see Docs https://reacttraining.com/reach-ui/tabs#tablist
 */
export const TabList = forwardRefWithAs<TabListProps, "div">(function TabList(
  { children, as: Comp = "div", onKeyDown, ...props },
  forwardedRef
) {
  const {
    isControlled,
    onSelectTab,
    onFocusPanel,
    setSelectedIndex,
    selectedIndex
  } = useTabsContext();

  const { descendants } = useContext(TabsDescendantsContext);

  let focusableTabs = useMemo(() => {
    let nodes: HTMLElement[] = [];
    for (let i = 0; i < descendants.length; i++) {
      let element = descendants[i].element;
      if (element && !boolOrBoolString(element.dataset.disabled)) {
        nodes.push(element);
      }
    }
    return nodes;
  }, [descendants]);

  let ownRef = useRef<HTMLElement | null>(null);
  let ref = useForkedRef(forwardedRef, ownRef);
  let isRTL = useRef(false);

  const getFocusIndices = useCallback(() => {
    /*
     * To detect RTL mode, we'll look first for <html dir="rtl">, as it is the
     * generally preferred method for handling RTL writing modes and is more
     * performant than checking computed styles (event though the computed
     * styles should take the dir attribute into account).
     */
    isRTL.current = false;
    if (
      ownRef.current &&
      ((ownRef.current.ownerDocument &&
        ownRef.current.ownerDocument.dir === "rtl") ||
        getStyle(ownRef.current, "direction") === "rtl")
    ) {
      isRTL.current = true;
    }

    /*
     * We could be clever and ~~functional~~ here but we really shouldn't need
     * to loop through these arrays more than once.
     *
     * TODO: We may want to check the document's active element here instead
     *       instead of the selectedIndex, even though you *shouldn't* be able
     *       to to focus a tab unless it's selected.
     */
    let selectedFocusIndex = focusableTabs.findIndex(
      element => element === descendants[selectedIndex].element
    );
    let first = -1;
    let last = -1;
    let prev = -1;
    let next = -1;
    for (let n = 0; n < descendants.length; n++) {
      let element = descendants[n].element;
      if (element === focusableTabs[0]) {
        first = n;
      }
      if (element === focusableTabs[focusableTabs.length - 1]) {
        last = n;
      }
      if (element === focusableTabs[selectedFocusIndex - 1]) {
        prev = n;
      }
      if (element === focusableTabs[selectedFocusIndex + 1]) {
        next = n;
      }
    }
    return {
      first,
      last,
      prev: prev === -1 ? last : prev,
      next: next === -1 ? first : next
    };
  }, [descendants, focusableTabs, selectedIndex]);

  // TODO: Determine proper behavior for Home/End key in RTL mode.
  function handleKeyDown(event: React.KeyboardEvent) {
    const { key } = event;

    // Bail if we aren't navigating
    if (
      !(
        key === "ArrowLeft" ||
        key === "ArrowRight" ||
        key === "ArrowDown" ||
        key === "Home" ||
        key === "End"
      )
    ) {
      return;
    }

    let { first, last, prev, next } = getFocusIndices();

    switch (key) {
      case "ArrowRight":
        onSelectTab(isRTL.current ? prev : next);
        break;
      case "ArrowLeft":
        onSelectTab(isRTL.current ? next : prev);
        break;
      case "ArrowDown":
        // don't scroll down
        event.preventDefault();
        onFocusPanel();
        break;
      case "Home":
        onSelectTab(first);
        break;
      case "End":
        onSelectTab(last);
        break;
      default:
        return;
    }
  }

  useLayoutEffect(() => {
    /*
     * In the event an uncontrolled component's selected index is disabled,
     * (this should only happen if the first tab is disabled and no default
     * index is set), we need to override the selection to the next selectable
     * index value.
     */
    if (
      !isControlled &&
      boolOrBoolString(descendants[selectedIndex]?.element?.dataset.disabled)
    ) {
      let { next } = getFocusIndices();
      setSelectedIndex(next);
    }
  }, [
    descendants,
    getFocusIndices,
    isControlled,
    selectedIndex,
    setSelectedIndex
  ]);

  return (
    <Comp
      role="tablist"
      {...props}
      data-reach-tab-list=""
      ref={ref}
      onKeyDown={wrapEvent(onKeyDown, handleKeyDown)}
    >
      {Children.map(children, (child, index) => {
        /*
         * TODO: Since refactoring to use context rather than depending on
         * parent/child relationships, we need to update our recommendations for
         * animations that break when we don't forward the `isSelected` prop
         * to our tabs. We will remove this in 1.0 and update our docs
         * accordingly.
         */
        return cloneValidElement(child, {
          isSelected: index === selectedIndex
        });
      })}
    </Comp>
  );
});

/**
 * @see Docs https://reacttraining.com/reach-ui/tabs#tablist-props
 */
export type TabListProps = {
  /**
   * `TabList` expects multiple `Tab` elements as children.
   *
   * `TabPanels` expects multiple `TabPanel` elements as children.
   *
   * @see Docs https://reacttraining.com/reach-ui/tabs#tablist-children
   */
  children?: React.ReactNode;
};

if (__DEV__) {
  TabList.displayName = "TabList";
  TabList.propTypes = {
    as: PropTypes.elementType,
    children: PropTypes.node
  };
}

////////////////////////////////////////////////////////////////////////////////

/**
 * Tab
 *
 * The interactive element that changes the selected panel.
 *
 * @see Docs https://reacttraining.com/reach-ui/tabs#tab
 */
export const Tab = forwardRefWithAs<
  // TODO: Remove this when cloneElement is removed
  TabProps & { isSelected?: boolean },
  "button"
>(function Tab(
  { children, isSelected: _, as: Comp = "button", disabled, ...props },
  forwardedRef
) {
  const {
    id: tabsId,
    onSelectTab,
    selectedIndex,
    userInteractedRef
  } = useTabsContext();
  const ownRef = useRef<HTMLElement | null>(null);
  const ref = useForkedRef(forwardedRef, ownRef);
  const index = useDescendant({
    element: ownRef.current,
    context: TabsDescendantsContext
  });

  const isSelected = index === selectedIndex;

  function onSelect() {
    onSelectTab(index);
  }

  useUpdateEffect(() => {
    if (isSelected && ownRef.current && userInteractedRef.current) {
      userInteractedRef.current = false;
      ownRef.current.focus();
    }
  }, [isSelected]);

  return (
    <Comp
      role="tab"
      {...props}
      ref={ref}
      data-reach-tab=""
      data-disabled={disabled}
      aria-controls={makeId(tabsId, "panel", index)}
      aria-disabled={disabled}
      aria-selected={isSelected}
      data-selected={isSelected ? "" : undefined}
      disabled={disabled}
      id={makeId(tabsId, "tab", index)}
      onClick={onSelect}
      tabIndex={isSelected ? 0 : -1}
    >
      {children}
    </Comp>
  );
});

/**
 * @see Docs https://reacttraining.com/reach-ui/tabs#tab-props
 */
export type TabProps = {
  disabled?: boolean;
} & TabPanelProps;

if (__DEV__) {
  Tab.displayName = "Tab";
  Tab.propTypes = {
    children: PropTypes.node,
    disabled: PropTypes.bool
  };
}

////////////////////////////////////////////////////////////////////////////////

/**
 * TabPanels
 *
 * The parent component of the panels.
 *
 * @see Docs https://reacttraining.com/reach-ui/tabs#tabpanels
 */
export const TabPanels = forwardRefWithAs<TabPanelsProps, "div">(
  function TabPanels({ children, as: Comp = "div", ...props }, forwardedRef) {
    let [tabPanels, setTabPanels] = useDescendants<HTMLElement>();
    return (
      <DescendantProvider
        context={TabPanelDescendantsContext}
        items={tabPanels}
        set={setTabPanels}
      >
        <Comp {...props} ref={forwardedRef} data-reach-tab-panels="">
          {children}
        </Comp>
      </DescendantProvider>
    );
  }
);

/**
 * @see Docs https://reacttraining.com/reach-ui/tabs#tabpanels-props
 */
export type TabPanelsProps = TabListProps & {};

if (__DEV__) {
  TabPanels.displayName = "TabPanels";
  TabPanels.propTypes = {
    as: PropTypes.elementType,
    children: PropTypes.node,
    selectedIndex: PropTypes.number
  };
}

////////////////////////////////////////////////////////////////////////////////

/**
 * TabPanel
 *
 * The panel that displays when it's corresponding tab is active.
 *
 * @see Docs https://reacttraining.com/reach-ui/tabs#tabpanel
 */
export const TabPanel = forwardRefWithAs<TabPanelProps, "div">(
  function TabPanel({ children, as: Comp = "div", ...props }, forwardedRef) {
    let { selectedPanelRef, selectedIndex, id: tabsId } = useTabsContext();
    let ownRef = useRef<HTMLElement | null>(null);

    let index = useDescendant({
      element: ownRef.current,
      context: TabPanelDescendantsContext
    });
    let isSelected = index === selectedIndex;

    let id = makeId(tabsId, "panel", index);

    let ref = useForkedRef(
      forwardedRef,
      ownRef,
      isSelected ? selectedPanelRef : null
    );

    return (
      <Comp
        hidden={!isSelected}
        role="tabpanel"
        {...props}
        ref={ref}
        data-reach-tab-panel=""
        aria-labelledby={makeId(tabsId, "tab", index)}
        id={id}
        tabIndex={-1}
      >
        {children}
      </Comp>
    );
  }
);

/**
 * @see Docs https://reacttraining.com/reach-ui/tabs#tabpanel-props
 */
export type TabPanelProps = {
  /**
   * `TabPanel` can receive any type of children.
   *
   * @see Docs https://reacttraining.com/reach-ui/tabs#tabpanel-children
   */
  children?: React.ReactNode;
};

if (__DEV__) {
  TabPanel.displayName = "TabPanel";
  TabPanel.propTypes = {
    as: PropTypes.elementType,
    children: PropTypes.node
  };
}

/**
 * Get a computed style value by property, backwards compatible with IE
 * @param element
 * @param styleProp
 */
function getStyle(element: HTMLElementWithCurrentStyle, styleProp: string) {
  let y: string | null = null;
  if (element.currentStyle) {
    y = element.currentStyle[styleProp];
  } else if (
    element.ownerDocument &&
    element.ownerDocument.defaultView &&
    typeof element.ownerDocument.defaultView.getComputedStyle === "function"
  ) {
    y = element.ownerDocument.defaultView
      .getComputedStyle(element, null)
      .getPropertyValue(styleProp);
  }
  return y;
}

////////////////////////////////////////////////////////////////////////////////
// Types

type HTMLElementWithCurrentStyle = HTMLElement & {
  currentStyle?: Record<string, string>;
};
