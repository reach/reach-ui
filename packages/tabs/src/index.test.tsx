import React from "react";
import { render, fireEvent } from "$test/utils";
import { Tabs, TabList, Tab, TabPanels, TabPanel } from "@reach/tabs";

describe("<Tabs />", () => {
  it("should match the snapshot", () => {
    const { asFragment } = render(<BasicTabs />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("focuses the correct tab with keyboard navigation", () => {
    const { getByText, getByRole, asFragment, container } = render(
      <BasicTabs />
    );
    const firstTab = getByText("Tab One");
    const tabList = getByRole("tablist");

    function getTabPanelByButtonClass(buttonId: string) {
      return container.querySelector(
        `#${container
          .querySelector(`.${buttonId}`)!
          .getAttribute("aria-controls")}`
      );
    }

    fireEvent.click(firstTab);

    fireEvent.keyDown(tabList, { key: "ArrowRight", code: 39 });
    expect(document.activeElement).toBe(container.querySelector(".tab-2"));
    expect(getTabPanelByButtonClass("tab-2")).toBeVisible();
    expect(getTabPanelByButtonClass("tab-1")).not.toBeVisible();
    expect(asFragment()).toMatchSnapshot();

    fireEvent.keyDown(tabList, { key: "ArrowRight", code: 39 });
    expect(document.activeElement).toBe(container.querySelector(".tab-3"));
    expect(getTabPanelByButtonClass("tab-3")).toBeVisible();
    expect(getTabPanelByButtonClass("tab-2")).not.toBeVisible();

    fireEvent.keyDown(tabList, { key: "ArrowRight", code: 39 });
    expect(document.activeElement).toBe(container.querySelector(".tab-1"));

    fireEvent.keyDown(tabList, { key: "ArrowLeft", code: 37 });
    expect(document.activeElement).toBe(container.querySelector(".tab-3"));

    fireEvent.keyDown(tabList, { key: "ArrowLeft", code: 37 });
    fireEvent.keyDown(tabList, { key: "ArrowLeft", code: 37 });
    expect(document.activeElement).toBe(container.querySelector(".tab-1"));

    fireEvent.keyDown(tabList, { key: "End", code: 35 });
    expect(document.activeElement).toBe(container.querySelector(".tab-3"));

    fireEvent.keyDown(tabList, { key: "Home", code: 36 });
    expect(document.activeElement).toBe(container.querySelector(".tab-1"));
  });
});

////////////////////////////////////////////////////////////////////////////////
function BasicTabs() {
  return (
    <div>
      <Tabs>
        <TabList>
          <Tab className="tab-1">Tab One</Tab>
          <Tab className="tab-2">Tab Two</Tab>
          <Tab className="tab-3">Tab Three</Tab>
        </TabList>

        <TabPanels>
          <TabPanel>
            <h1>one!</h1>
            <button>yo</button>
          </TabPanel>
          <TabPanel>
            <h1>two!</h1>
          </TabPanel>
          <TabPanel>
            <h1>three!</h1>
          </TabPanel>
        </TabPanels>
      </Tabs>
    </div>
  );
}
