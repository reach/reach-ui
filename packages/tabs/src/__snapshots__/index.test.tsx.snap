// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> focuses the correct tab with keyboard navigation 1`] = `
<DocumentFragment>
  <div>
    <div
      data-reach-tabs=""
    >
      <div
        data-reach-tab-list=""
        role="tablist"
      >
        <button
          aria-controls="tabs--4--panel--0"
          aria-selected="false"
          class="tab-1"
          data-reach-tab=""
          id="tabs--4--tab--0"
          role="tab"
          tabindex="-1"
        >
          Tab One
        </button>
        <button
          aria-controls="tabs--4--panel--1"
          aria-selected="true"
          class="tab-2"
          data-reach-tab=""
          data-selected=""
          id="tabs--4--tab--1"
          role="tab"
          tabindex="0"
        >
          Tab Two
        </button>
        <button
          aria-controls="tabs--4--panel--2"
          aria-selected="false"
          class="tab-3"
          data-reach-tab=""
          id="tabs--4--tab--2"
          role="tab"
          tabindex="-1"
        >
          Tab Three
        </button>
      </div>
      <div
        data-reach-tab-panels=""
      >
        <div
          aria-labelledby="tabs--4--tab--0"
          data-reach-tab-panel=""
          hidden=""
          id="tabs--4--panel--0"
          role="tabpanel"
          tabindex="-1"
        >
          <h1>
            one!
          </h1>
          <button>
            yo
          </button>
        </div>
        <div
          aria-labelledby="tabs--4--tab--1"
          data-reach-tab-panel=""
          id="tabs--4--panel--1"
          role="tabpanel"
          tabindex="-1"
        >
          <h1>
            two!
          </h1>
        </div>
        <div
          aria-labelledby="tabs--4--tab--2"
          data-reach-tab-panel=""
          hidden=""
          id="tabs--4--panel--2"
          role="tabpanel"
          tabindex="-1"
        >
          <h1>
            three!
          </h1>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Tabs /> should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      data-reach-tabs=""
    >
      <div
        data-reach-tab-list=""
        role="tablist"
      >
        <button
          aria-controls="tabs--1--panel--0"
          aria-selected="true"
          class="tab-1"
          data-reach-tab=""
          data-selected=""
          id="tabs--1--tab--0"
          role="tab"
          tabindex="0"
        >
          Tab One
        </button>
        <button
          aria-controls="tabs--1--panel--1"
          aria-selected="false"
          class="tab-2"
          data-reach-tab=""
          id="tabs--1--tab--1"
          role="tab"
          tabindex="-1"
        >
          Tab Two
        </button>
        <button
          aria-controls="tabs--1--panel--2"
          aria-selected="false"
          class="tab-3"
          data-reach-tab=""
          id="tabs--1--tab--2"
          role="tab"
          tabindex="-1"
        >
          Tab Three
        </button>
      </div>
      <div
        data-reach-tab-panels=""
      >
        <div
          aria-labelledby="tabs--1--tab--0"
          data-reach-tab-panel=""
          id="tabs--1--panel--0"
          role="tabpanel"
          tabindex="-1"
        >
          <h1>
            one!
          </h1>
          <button>
            yo
          </button>
        </div>
        <div
          aria-labelledby="tabs--1--tab--1"
          data-reach-tab-panel=""
          hidden=""
          id="tabs--1--panel--1"
          role="tabpanel"
          tabindex="-1"
        >
          <h1>
            two!
          </h1>
        </div>
        <div
          aria-labelledby="tabs--1--tab--2"
          data-reach-tab-panel=""
          hidden=""
          id="tabs--1--panel--2"
          role="tabpanel"
          tabindex="-1"
        >
          <h1>
            three!
          </h1>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
